### DISCUSSION TITLE
Having issues implementing FromStr for a struct

### DISCUSSION THREAD

**Discussion by djacu** [winnow-rs/winnow#639] [Q&A]
## Background context

I am building my own parser to parse through Nix derivation (.drv) files. Source for the repository and the branch I'm currently working on can be found here [djacu/nix-derivation-tools](https://github.com/djacu/nix-derivation-tools/blob/switch-from-nom-to-minnow/packages/nix-derivation-parser/).

The exact file where I am testing things out is here: https://github.com/djacu/nix-derivation-tools/blob/switch-from-nom-to-minnow/packages/nix-derivation-parser/src/derivations/parsers.rs

## Issue

I read through the tutorials and the [Implementing `FromStr`](https://docs.rs/winnow/0.6.20/winnow/_topic/fromstr/index.html) topic and wanted to do the same for my project. Specifically on this [`Derivation`](https://github.com/djacu/nix-derivation-tools/blob/fcfca0c5b2e529380e6b3a6b7aa3e6f28ebd4614/packages/nix-derivation-parser/src/derivations/types.rs#L20) struct.

I started with something similar to the topic like this:

```rust
impl std::str::FromStr for Derivation {
    type Err = String;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        parse_derivation::<Derivation>.parse(input).map_err(|e| e.to_string())
    }
}
```

but was getting error like this

> the method `to_string` exists for struct `ParseError<&str, Derivation>`, but its trait bounds were not satisfied  the following trait bounds were not satisfied:  `ParseError<&str, Derivation>: std::fmt::Display`  which is required by `ParseError<&str, Derivation>: ToString`

I tried implementing a dummy `Display` to see if that would satisfy the trait requirements like so:

```rust
impl std::fmt::Display for Derivation {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
       write!(f, "{}", "") 
    }
}
```

but that produced an even more inscrutable error.

> the trait bound `Derivation: ParserError<&str>` is not satisfied  the following other types implement trait `ParserError<I>`:    ()    ContextError<C>    ErrMode<E>    InputError<I>    TreeError<I, C>    winnow::error::ErrorKind

What am I doing wrong?

**Comment by epage** [comment]
> `parse_derivation::<Derivation>.parse(input)`

`Derivation` is your Output type but you are passing it to your parser as the Error type.  Instead, select a different error, like `ContextError`.

```rust
impl std::str::FromStr for Derivation {
    type Err = String;

    fn from_str(input: &str) -> Result<Self, Self::Err> {
        parse_derivation::<ContextError>.parse(input).map_err(|e| e.to_string())
    }
}
```
Unless you are building reusable parsers  for others, I'd recommend hard coding the Error type to the type that you prefer, much like you hard code the Input type.  Note that `ContextError` is the default for `PResult`
